Gutenberg: Add Top Posts and Pages Block

Aurorum opened the pull request:
Fixes #32937

## Proposed changes:

Converts the legacy Top Posts & Pages widget into a Gutenberg block! It has the same features; you can still choose to filter out different types of media and customise the display, but the block also allows you to choose the timeframe of stats which to draw on. 

The only exception is that this block is based entirely on views/traffic. The widget allowed determining top content from Likes, but I don't have access to a WordPress.com environment that would allow me to add this, so I'm afraid that somebody else would need to handle that in future. I personally don't feel like that's a massive deal though, at least not for this first version. 

<img width="1580" alt="Screenshot 2023-11-17 at 23 34 23" src="https://github.com/Automattic/jetpack/assets/43215253/d15a1334-cbd3-4509-8858-5483d98cad77">

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
N/A

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
N/A

## Testing instructions:
I'm hoping that the block should be fairly intuitive to use! Just like the Related Posts block, you should be prompted to activate Stats if the module is disabled. In fact, I've drawn quite heavily on that block - for example, the metadata and layout settings are the same. 

<img width="761" alt="Screenshot 2023-11-18 at 15 26 04" src="https://github.com/Automattic/jetpack/assets/43215253/a867fff5-cd8f-43cd-a113-f2255681c7cb">

If Stats is active, some posts should automatically display. If there aren't enough posts within the given timeframe, then random posts should be used to fill the rest of the block. There are some settings to filter out content and to choose which metadata information to display, so definitely verify that those work. 

Ultimately though, it should render nicely in the frontend. 

<img width="697" alt="Screenshot 2023-11-18 at 15 28 42" src="https://github.com/Automattic/jetpack/assets/43215253/2264803f-63dd-44e8-aac8-36de236dd4b2">

Aurorum committed:
WIP

Aurorum committed:
WIP

Aurorum committed:
WIP

Aurorum committed:
Merge branch 'add/top-post-pages-block' of https://github.com/Aurorum/jetpack into add/top-post-pages-block

Aurorum committed:
WIP

Aurorum committed:
Merge branch 'trunk' into add/top-post-pages-block

Aurorum committed:
WIP

Aurorum committed:
WIP

Aurorum committed:
Changelog

Aurorum commented:
This has been a very fun way to spend the week! Should be ready for review now - no doubt that there'll still be a ton of issues, but it seems to work!! :)

Aurorum committed:
Add missing tests

Aurorum committed:
Add changelog to fix PR label

jeherve commented:
Thanks for your work so far! I've allowed the tests to run so you can get a first round of automated feedback on your work. Do you think you could look at the test failures?

Aurorum committed:
Address test feedback

Aurorum commented:
Thanks @jeherve! Should be all fixed now - would you mind running them again?

Aurorum committed:
Attempt 2 at tests fix

Aurorum commented:
Looks like they caused new errors - those should (hopefully!) be fixed now 🤞

Aurorum committed:
Tests fix v3

Aurorum committed:
Add controls tests

Aurorum commented:
Sorry @jeherve - just added some more tests for the Block controls which I missed (they're probably important since controls can vary based on post types). Would you remind re-triggering tests again when you get the chance? Thanks!

Aurorum committed:
Remove unneeded function

Aurorum committed:
Fix tests v4

Aurorum commented:
It's like whack-a-mole; fix one error, and three new ones appear 😭 Thanks to whoever ran them again - I think that I've committed the final fixes now!

Aurorum committed:
Merge remote-tracking branch 'upstream/trunk' into add/top-post-pages-block

Aurorum committed:
Fix tests v5; update styling

Aurorum commented:
I've synced with `trunk` to ensure the PR is up-to-date, fixed the final whitespace issues, and updated some of the styling so that it's identical to how Related Posts are handled. In summary, this means having two items fill one row when there's five or seven items, like so. It also means having a List View on smaller screens rather than narrowing the size of the grid. I think that's a sensible precedent to follow. 

<img width="786" alt="Screenshot 2023-11-22 at 22 20 39" src="https://github.com/Automattic/jetpack/assets/43215253/c88da666-51e2-446a-8287-6854948bb171">

@jeherve I think that should cover the `Needs Author Reply`, but please let me know if there's anything else which I need to take care of for now!

jeherve commented:
I'm afraid the tests are still failing at the moment (see below). Do you think you could take a look at them?

Thank you!

Aurorum commented:
@jeherve I’m afraid that I’m a little confused by them - does Jetpack no longer use `$$next-version$$`? I don’t quite understand why it should be causing an issue.

`plugins/jetpack 2:24.992] ::error file=projects/plugins/jetpack/extensions/blocks/top-posts/block.json,line=8::Unexpected $$next-version$$ token.`

jeherve commented:
> @jeherve I’m afraid that I’m a little confused by them - does Jetpack no longer use `$$next-version$$`? I don’t quite understand why it should be causing an issue.
> 
> `plugins/jetpack 2:24.992] ::error file=projects/plugins/jetpack/extensions/blocks/top-posts/block.json,line=8::Unexpected $$next-version$$ token.`

This specific error is something we may need to fix on our end I think, that's a good catch! cc @anomiex Do we need to update the script to take the new `block.json` file structure into account?

Aurorum commented:
I think that `Tests / PHP tests` errors are unrelated to these changes based on a browse through them - please let me know if I’m missing something though. 

![72F81C8B-75C8-4AC9-A298-AB3BCBF37D49](https://github.com/Automattic/jetpack/assets/43215253/9b0f65a9-52f7-41e3-bbc4-f65961ee6c44)

jeherve commented:
> I think that `Tests / PHP tests` errors are unrelated to these changes based on a browse through them - please let me know if I’m missing something though.
> 
> ![72F81C8B-75C8-4AC9-A298-AB3BCBF37D49](https://user-images.githubusercontent.com/43215253/285508898-9b0f65a9-52f7-41e3-bbc4-f65961ee6c44.jpeg)

This is the key part of the tests that would need to be addressed:

> The REST API route definition for <code>wpcom/v2/top-posts</code> is missing the required <code>permission_callback</code> argument. For REST API routes that are intended to be public, use <code>__return_true</code> as the permission callback. (This message was added in version 5.5.0.)

Aurorum committed:
Add permission_callback to endpoint

Aurorum commented:
Thanks! I've added a `permission_callback` to that endpoint - would you mind running the tests again?

Aurorum commented:
Could I just confirm that the remaining two failures aren’t down to these changes? I’m happy to switch `$$next-version$$` to a hard-coded number instead, but not sure if that would actually be helpful here.

Edit: I ran `tools/project-version.sh` as the tests advised; it seems to want to me update the `composer.json` files, which I've done. This is a little out of my depth though, so let me know if I'm misunderstanding.

Aurorum committed:
Run tools/project-version.sh

Aurorum committed:
Run tools/project-version.sh

Aurorum committed:
Fix links

anomiex commented:
> This specific error is something we may need to fix on our end I think, that's a good catch! cc @anomiex Do we need to update the script to take the new `block.json` file structure into account?

The [wporg docs for `version`](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#version) say it's supposed to be the "version of the block". Does putting in the version number of the plugin actually make sense for that? Will we actually remember to update it in each block's `block.json` when an update is needed?

Aurorum committed:
Merge remote-tracking branch 'upstream/trunk' into add/top-post-pages-block

Aurorum committed:
Ensure PR is up to date again

Aurorum committed:
Update index.json

Aurorum committed:
Update block.json version number

Aurorum commented:
> The [wporg docs for version](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#version) say it's supposed to be the "version of the block". Does putting in the version number of the plugin actually make sense for that? Will we actually remember to update it in each block's block.json when an update is needed?

With that in mind, I've set the version to `1.0` in-line with that documentation. I'm happy to change it afterwards (please let me know if you'd prefer that!), but it should mean that all tests successfully pass in the meantime.

Aurorum committed:
Run tools/composer-update-monorepo.sh

Aurorum commented:
I just pushed another commit to handle the final issue of an outdated package - should be the last time that tests need to be re-run!! Thank you. :)

Aurorum commented:
Looks like the tests are finally passing (!!) @jeherve - please let me know if there’s anything else which I need to cover as part of the Needs Author Reply label!

kangzj reviewed (dismissed):
Thanks for the great work @Aurorum 👍 

I had a go and it looks awesome and I'm happy for it to merge as a beta block. I'll loop in @a8ck3n who worked on related posts before for more reviews.

Some improvement nitpick:

<img width="646" alt="Screenshot 2023-11-29 at 11 41 47 AM" src="https://github.com/Automattic/jetpack/assets/1425433/4c3e8c04-cf7a-4af2-9dcf-abab3be08a0e">


(1) It'll be great to have different style for title/date etc like related posts.

(2) I would love it more if we have place holders for posts that don't have thumbnails.

I don't think either of them are blockers.

Aurorum commented:
Thanks so much @kangzj!! Totally agree, and happy to add both those improvements, either here or a future PR - whichever you’d prefer. 

> It'll be great to have different style for title/date etc like related posts

For this, do you mean the typography settings? I couldn’t see anything different from Related Posts, but it does feel a little odd when the heading is large and this text remains quite small.

Thanks again for the review! :)

Aurorum committed:
Add placeholder when missing thumbnails

Aurorum commented:
I've added some improvements for posts without thumbnails: it now gets an image from the post if there's no Featured Image set. I think that's what Publicize does, so hopefully that seems reasonable to follow. 

When there's still no Featured Image and there aren't any images within the post to draw upon, I've added a grey placeholder so that things looks less out of place. It's hidden on mobile where there's no grid. 

<img width="773" alt="Screenshot 2023-11-29 at 17 27 15" src="https://github.com/Automattic/jetpack/assets/43215253/6331cf70-836b-4e41-a718-78187cf49712">

Let me know whether that's what you had in mind, or if there's a better way of doing it!

a8ck3n commented:
> I had a go and it looks awesome and I'm happy for it to merge as a beta block. I'll loop in @a8ck3n who worked on related posts before for more reviews.

I'm having a few issues with my dev environment but once I get that sorted I'll give things a look.

cc: @Aurorum

Aurorum committed:
Merge remote-tracking branch 'upstream/trunk' into add/top-post-pages-block

kangzj reviewed (dismissed):
Sorry I was tied up a bit last week. I'll have another spin tomorrow and merge it when appropriate. Thanks again for your contribution and patience with us @Aurorum 🙏

Aurorum commented:
No worries, thanks so much @kangzj! Feel free to let me know if there’s anything that needs changing (either before or after this PR is merged; happy to make sure it’s completely up to scratch).

kangzj commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-top-posts.php:
```diff
@@ -0,0 +1,201 @@
+		);
+
+		$data            = ( new WPCOM_Stats() )->get_top_posts( $query_args, $override_cache );
+		$posts_retrieved = count( $data['summary']['postviews'] );
```
It's possible that the $data is an instance of WP_Error, so it'll be great if we could give it a check and resolve the flow here, for example:

```suggestion
		if ( is_wp_error( $data ) ) {
			$data['summary']['postviews'] = array();
		}
		$posts_retrieved = count( $data['summary']['postviews'] );
```

Aurorum replied:
Added!

kangzj replied:
Hi @Aurorum Thanks for addressing it so quickly. My bad, `$data `is still an object there and probably we should do this:

```PHP
$data = array( 'summary' => array( 'postviews' => array() ) );
```

Aurorum replied:
Whoops, fixed!

kangzj commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-top-posts.php:
```diff
@@ -0,0 +1,201 @@
+<?php
```
The endpoint needs to be wired up for WPCOM, which I can help with after the PR is merged.

kangzj commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-top-posts.php:
```diff
@@ -0,0 +1,201 @@
+			'period'    => 'day',
+		);
+
+		$data            = ( new WPCOM_Stats() )->get_top_posts( $query_args, $override_cache );
```
Hi @jeherve,

Just in case you have some insights. When I tried to test the endpoint on WPCOM, the call constantly fails. I had look. On WPCOM, the API call was rerouted to use `WPCOM_API_Direct::do_request()`, and then fails with an error: 
```
{"code":"invalid_namespace","message":"Invalid API namespace","data":{"status":404}}
```

I searched the error message, which seems to be thrown by `wp-content/rest-api-plugins/centralized/centralize.php` - `wpcom/v2`. I'm wondering why `v1` endpoints couldn't be found 🤔

jeherve replied:
> I'm wondering why v1 endpoints couldn't be found 🤔

Since the diff is outdated I'm not sure I fully understand, but I don't believe we ever had `wpcom/v1` endpoints in prod.

kangzj commented on projects/plugins/jetpack/extensions/blocks/top-posts/top-posts.php:
```diff
@@ -0,0 +1,104 @@
+	 * top posts might have changed since then. As such, we must
+	 * make another request to check for updated stats.
+	 */
+	$request_url = sprintf(
```
The request here wouldn't work on WPCOM, as `wpcom/v2` would be loaded to `public-api.wordpress.com`, so the API wouldn't be first party anymore.

I'm wondering whether the remote request could be saved by calling a function that returns top posts, which could save us the HTTP request overhead, and hence improve the performance of the block. More importantly, it would work with WP.COM with no issues. To accomplish this, we could refactor the following code to a helper or something be shared by the API and the block.

https://github.com/Automattic/jetpack/pull/34153/files#diff-b82973540f31245c9564a0c3e997adba550cc9387fa7249be99581128844b45eR106

Aurorum replied:
Ooh, yes, that makes sense. Since we've added a check here to remove the block from WP.com, should I leave it in this PR for now? I'll definitely refactor the code in a new one if so. :)

kangzj replied:
Yes, we can leave it for now 😄

Aurorum replied:
PR for this in #34479 :)

kangzj commented on projects/plugins/jetpack/extensions/blocks/top-posts/top-posts.php:
```diff
@@ -0,0 +1,104 @@
+ * registration if we need to.
+ */
+function register_block() {
+	Blocks::jetpack_register_block(
+		__DIR__,
+		array( 'render_callback' => __NAMESPACE__ . '\load_assets' )
+	);
```
```suggestion
        if ( ! defined( 'IS_WPCOM' ) && ( ( new Connection_Manager( 'jetpack' ) )->has_connected_owner() && ! ( new Status() )->is_offline_mode() ) )
	Blocks::jetpack_register_block(
		__DIR__,
		array( 'render_callback' => __NAMESPACE__ . '\load_assets' )
	);
```

We'll need the following classes:

```PHP
use Automattic\Jetpack\Connection\Manager as Connection_Manager;
use Automattic\Jetpack\Status;
```

Aurorum replied:
Checks added - thank you!

kangzj reviewed (commented):
Hi @Aurorum,

Great work! I reviewed all the code, looking awesome 👍

The code is very thorough and enjoyable to read! And thank you so much for adding all the tests 🏆 

I only have two feedback before merging 😄 

### 1. Error handling

https://github.com/Automattic/jetpack/pull/34153/files#diff-b82973540f31245c9564a0c3e997adba550cc9387fa7249be99581128844b45eR127

### 2. Register the block only for Jetpack

As we still need to figure out a couple of things for WP.COM, could we please restrict the block for Jetpack only at the stage? And we could fix it in follow-up PRs and then enable the block for WP.COM.

https://github.com/Automattic/jetpack/pull/34153/files#diff-7a5ddadbca141cf1cea2c9f32c5ba878c8369d633168ccae7608a89ccdaac71dR20

Thanks again for your contribution. We really appreciate it 🤝 👍

kangzj commented:
Another nitpick for styling in editor:

![image](https://github.com/Automattic/jetpack/assets/1425433/d370874b-1f40-4434-93c3-71093ae444b7)

![image](https://github.com/Automattic/jetpack/assets/1425433/059de528-54d4-48cf-b0a4-c152e077c83e)

Should we align the block from the top instead?

It seems to be working okay with images from the frontend:

![image](https://github.com/Automattic/jetpack/assets/1425433/7589d9ed-cc6f-44b5-9a3b-16841f8501b5)

But not only titles:
![image](https://github.com/Automattic/jetpack/assets/1425433/2c025a16-3ada-42a3-bf7d-9b036bc77cd0)

Thanks again 🙇

Aurorum committed:
Address review

Aurorum commented:
Thanks so much for your detailed review @kangzj - very kind of you! :) I've addressed the error handling, unregistering for WP.com and alignment issues - totally agree that it looks weird starting from the bottom.

Aurorum committed:
Fix error handling

kangzj commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-top-posts.php:
```diff
@@ -0,0 +1,206 @@
+			$thumbnail = get_the_post_thumbnail_url( $post_id );
+
+			if ( ! $thumbnail ) {
+				$post_image = reset( get_attached_media( 'image', $post_id ) );
+				$thumbnail  = wp_get_attachment_url( $post_image->ID );
```
```suggestion
				$post_images = get_attached_media( 'image', $post_id );
				$post_image  = reset( $post_images );
				if ( $post_image ) {
					$thumbnail = wp_get_attachment_url( $post_image->ID );
				}
```

Discovered two warning here:

1. reset requires a variable
2. reset could return false if the array is empty

kangzj commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-top-posts.php:
```diff
@@ -0,0 +1,206 @@
+					'title'     => $post['title'],
+					'type'      => $post['type'],
+					'public'    => $post['public'],
+					'views'     => $post['views'],
```
```suggestion
					'views'     => isset( $post['views'] ) ? $post['views'] : 0,
```

Address an warning here too.

kangzj reviewed (commented):
Thanks for addressing the feedback so quickly. I promise these are the last two tweaks we need to merge it 😄 Thanks for your patience 🤝

kangzj committed:
Update projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-top-posts.php

kangzj committed:
Update projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-top-posts.php

kangzj reviewed (approved):
I went ahead and committed the couple of little tweaks. Hope you don't mind. Let's :shipit: 🎉 🎉 🎉

Aurorum commented:
Not a problem at all, thank you so much for helping get this over the line!!

jeherve reviewed (commented):
Nice work on this one! 👏 

I'm catching up on this with a follow-up note. Since the block relies on the Stats module, we would need to only make the full block available when the Stats module is enabled, and offer a way to turn on the module from the block settings when it is inactive.
You can check the different Pull Requests in #32401 to figure out examples of how this can be done.

I opened #34484 to keep track of that.
